File: Bioinformatics Stronghold/005_Computing_GC_Content.pl
#!/usr/bin/perl
use warnings;
use strict;

my $fileName = shift or die "No file provided\n";

# Read sequences from fasta file
my %sequences = %{readFasta($fileName)};

# Retrieve sequence with highest GC%
my $topHeader;
my $topGC = 0;
foreach(keys(%sequences)) {
    my $GC = contentGC($sequences{$_}, 1, 1);
    if($GC > $topGC) {
        $topHeader = $_;
        $topGC = $GC;
    }
}

printf("%s\n%f\n", $topHeader, $topGC);

exit;

sub readFasta{
    my ($fileName) = @_;
    my %sequences;
    my $header;

    open(F, $fileName) or die "File not found\n";

    while(<F>) {
        chomp($_);
        next if($_ =~ /^\s*$/); # Skip empty lines
        if($_ =~ /^>/) {
            $header = $_;
            $header =~ s/>//;
            if (exists $sequences{$header}) {
                die "Duplicated entries in file: $header\n";
            }
        } else {
            next if(!$header);
            $sequences{$header} .= $_; # Concatenate sequence fragments
        }
    }
    close(F);
    return(\%sequences);
}

sub contentGC {
    my ($seq, $step, $window) = @_;
    my @GC;

    if($step > $window) {
        die "Step size cannot be greater than window\n";
    }

    # Whole sequence %GC with step = 1 & window = 1
    if($step == 1 && $window == 1) {
        return(($seq =~ tr/GC//) / length($seq) * 100);
    
    # Otherwise returns array with sliding window GC%
    } else {
        for (my $i = 0; $i < length($seq); $i += $step) {
            my $frag = substr($seq, $i, $window);
            push(@GC, ($frag =~ tr/GC//) / $window * 100)
        }
        return(\@GC)
    }
}

File: Bioinformatics Stronghold/005_Computing_GC_Content.py
#!/usr/bin/env python3

import re
from sys import exit, argv

def main():
    if argv[1]:
        fileName = argv[1]
    else:
        exit("No file provided")
    
    # Read sequences from fasta file
    sequences = readFasta(fileName)

    # Retrieve sequence with highest GC%
    topGC = 0
    for key in sequences.keys():
        GC = contentGC(sequences[key], 1, 1)
        if GC > topGC:
            topHeader = key
            topGC = GC

    print("{}\n{:.6f}".format(topHeader, topGC))


def readFasta(fileName):
    sequences = {}

    with open(fileName, 'r') as F:
        seq = []
        for line in F:

            if re.match(r'^\s*$', line):
                continue

            if line.startswith('>'):
                if seq:
                    sequences[header] = ''.join(seq)
                seq = []
                header = line.strip()[1:]
                if header in sequences:
                    exit("Duplicated entries in file: {}".format(header))
            else:
                if not header:
                    continue
                seq.append(line.strip().upper())
        
        # Store last sequence in file
        sequences[header] = ''.join(seq)
    
    return(sequences)

def contentGC(seq, step, window):
    if step > window:
        exit("Step size cannot be greater than window")
    
    # Whole sequence %GC with step = 1 & window = 1
    if step == 1 & window == 1:
        return(len(re.findall(r'[GC]', seq)) / len(seq) * 100)
    
    # Otherwise returns array with sliding window GC%
    else:
        GC = []
        for i in range(0, len(seq), step):
            frag = seq[i:i+step]
            GC.append(len(re.findall(r'[GC]', seq)) / len(seq) * 100)
        return(GC)
        

if __name__ == '__main__':
    main()